### PR TITLE
[ml] improve managed network schema

### PR DESCRIPTION
**Pull request by joshharrin**
# Description

Work item: https://dev.azure.com/msdata/Vienna/_workitems/edit/2475216/

Workspace managed network related schema is correct and behaving as intended but the tool used to generate latest yaml schema does not correctly handle Raw marshmallow fields. To fix this, adjust the managed network schema to use unionfield so that generated json schema is as intended. This also can help make it slightly cleaner. Generated schema [azuremlschemas.azureedge.net/latest/workspace.schema.json](https://azuremlschemas.azureedge.net/latest/workspace.schema.json) will then be up to date at the next release of SDK

Tests:
using yaml schema for workspace with all type of outbound rule, for example below, ensure that load from yaml to workspace object and dump to yaml functionality continues to behave as expected. 

YAML file:
image

test and output
image

Confirmed new generated workspace json schema will be as intended using _**python .\scripts\json_schema_generator.py**_ from sdk-cli-v2 repo:
```yaml
        "ManagedNetworkSchema": {
            "type": "object",
            "properties": {
                "isolation_mode": {
                    "type": "string",
                    "enum": [
                        "disabled",
                        "allow_internet_outbound",
                        "allow_only_approved_outbound"
                    ],
                    "title": "isolation_mode"
                },
                "network_id": {
                    "title": "network_id",
                    "type": "string",
                    "readonly": true
                },
                "outbound_rules": {
                    "title": "outbound_rules",
                    "type": [
                        "array",
                        "null"
                    ],
                    "items": {
                        "oneOf": [
                            {
                                "type": "object",
                                "$ref": "#/definitions/PrivateEndpointOutboundRuleSchema"
                            },
                            {
                                "type": "object",
                                "$ref": "#/definitions/ServiceTagOutboundRuleSchema"
                            },
                            {
                                "type": "object",
                                "$ref": "#/definitions/FqdnOutboundRuleSchema"
                            }
                        ]
                    }
                },
                "status": {
                    "type": "object",
                    "$ref": "#/definitions/ManagedNetworkStatusSchema"
                }
            },
            "additionalProperties": false
        },
        "PrivateEndpointOutboundRuleSchema": {
            "type": "object",
            "required": [
                "destination",
                "name"
            ],
            "properties": {
                "category": {
                    "type": "string",
                    "enum": [
                        "required",
                        "recommended",
                        "user_defined"
                    ],
                    "title": "category",
                    "readonly": true
                },
                "destination": {
                    "type": "object",
                    "$ref": "#/definitions/PrivateEndpointDestinationSchema"
                },
                "name": {
                    "title": "name",
                    "type": "string"
                },
                "status": {
                    "title": "status",
                    "type": "string",
                    "readonly": true
                },
                "type": {
                    "const": "private_endpoint"
                }
            },
            "additionalProperties": false
        },
        "PrivateEndpointDestinationSchema": {
            "type": "object",
            "required": [
                "service_resource_id",
                "spark_enabled",
                "subresource_target"
            ],
            "properties": {
                "service_resource_id": {
                    "title": "service_resource_id",
                    "type": "string"
                },
                "spark_enabled": {
                    "title": "spark_enabled",
                    "type": "boolean"
                },
                "subresource_target": {
                    "title": "subresource_target",
                    "type": "string"
                }
            },
            "additionalProperties": false
        },
        "ServiceTagOutboundRuleSchema": {
            "type": "object",
            "required": [
                "destination",
                "name"
            ],
            "properties": {
                "category": {
                    "type": "string",
                    "enum": [
                        "required",
                        "recommended",
                        "user_defined"
                    ],
                    "title": "category",
                    "readonly": true
                },
                "destination": {
                    "type": "object",
                    "$ref": "#/definitions/ServiceTagDestinationSchema"
                },
                "name": {
                    "title": "name",
                    "type": "string"
                },
                "status": {
                    "title": "status",
                    "type": "string",
                    "readonly": true
                },
                "type": {
                    "const": "service_tag"
                }
            },
            "additionalProperties": false
        },
        "ServiceTagDestinationSchema": {
            "type": "object",
            "required": [
                "port_ranges",
                "protocol",
                "service_tag"
            ],
            "properties": {
                "port_ranges": {
                    "title": "port_ranges",
                    "type": "string"
                },
                "protocol": {
                    "title": "protocol",
                    "type": "string"
                },
                "service_tag": {
                    "title": "service_tag",
                    "type": "string"
                }
            },
            "additionalProperties": false
        },
        "FqdnOutboundRuleSchema": {
            "type": "object",
            "required": [
                "destination",
                "name"
            ],
            "properties": {
                "category": {
                    "type": "string",
                    "enum": [
                        "required",
                        "recommended",
                        "user_defined"
                    ],
                    "title": "category",
                    "readonly": true
                },
                "destination": {
                    "title": "destination",
                    "type": "string"
                },
                "name": {
                    "title": "name",
                    "type": "string"
                },
                "status": {
                    "title": "status",
                    "type": "string",
                    "readonly": true
                },
                "type": {
                    "const": "fqdn"
                }
            },
            "additionalProperties": false
        },
        "ManagedNetworkStatusSchema": {
            "type": "object",
            "properties": {
                "spark_ready": {
                    "title": "spark_ready",
                    "type": "boolean",
                    "readonly": true
                },
                "status": {
                    "title": "status",
                    "type": "string",
                    "readonly": true
                }
            },
            "additionalProperties": false
        },
```

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md##building-and-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

**Review by kwucode — APPROVED** [review]

File: sdk/ml/azure-ai-ml/azure/ai/ml/_schema/workspace/networking.py [hunk 1]
```diff
@@ -125,9 +182,19 @@ class ManagedNetworkSchema(metaclass=PatchedSchemaMeta):
...
-    network_id = fields.Str(required=False)
-    status = NestedField(ManagedNetworkStatusSchema, dump_only=True)
+    outbound_rules = fields.List(
+        UnionField(
+            [
+                NestedField(FqdnOutboundRuleSchema, allow_none=False, unknown=EXCLUDE),
+                NestedField(ServiceTagOutboundRuleSchema, allow_none=False, unknown=EXCLUDE),
+                NestedField(PrivateEndpointOutboundRuleSchema, allow_none=False, unknown=EXCLUDE),
+            ],
+            allow_none=False,
```
Comment (nthandeMS) [inline sdk/ml/azure-ai-ml/azure/ai/ml/_schema/workspace/networking.py:192]: The schema will generate Union Fields with the `anyOf` by default. If the user can only provide a YAML matching one of these nested schemas, you can use `oneOf` instead: 
```suggestion
            allow_none=False,
            is_strict=True,
```

**Review by nthandeMS — APPROVED** [review]

File: sdk/ml/azure-ai-ml/azure/ai/ml/_schema/workspace/networking.py (hunk 1 above)
Comment (nthandeMS) [reply sdk/ml/azure-ai-ml/azure/ai/ml/_schema/workspace/networking.py:192]: This also has the added benefit of looking better in intellisense when using VS Code.

File: sdk/ml/azure-ai-ml/azure/ai/ml/_schema/workspace/networking.py [hunk 2]
```diff
@@ -114,6 +114,63 @@ def service_tag_dest2dict(self, service_tag, protocol, port_ranges):
...
+@experimental
+class PrivateEndpointOutboundRuleSchema(metaclass=PatchedSchemaMeta):
+    name = fields.Str(required=True)
+    type = StringTransformedEnum(
+        allowed_values=[
+            OutboundRuleType.PRIVATE_ENDPOINT,  # use allowed values to differentiate for unionfield ordering
+        ],
+        casing_transform=camel_to_snake,
+        metadata={"description": "outbound rule type."},
+    )
```
Comment (seanyao1) [inline sdk/ml/azure-ai-ml/azure/ai/ml/_schema/workspace/networking.py:133]: Shouldn't type be a constant for this class?

Comment (joshharrin) [reply sdk/ml/azure-ai-ml/azure/ai/ml/_schema/workspace/networking.py]: applied, thanks. was using enum with only one option to properly match type used by UnionField. Tested with marshmallow constant field and that can work as long as order for unionfields is correct. applied

File: sdk/ml/azure-ai-ml/azure/ai/ml/_schema/workspace/networking.py (hunk 1 above)
Comment (joshharrin) [reply sdk/ml/azure-ai-ml/azure/ai/ml/_schema/workspace/networking.py:179]: applied, thanks!

**Review by seanyao1 — APPROVED** [review]

**Pull request merged**